Build: Support yarn for installing .hintrc packages

shivangg opened the pull request:
<!--

Read our pull request guide:
https://webhint.io/docs/contributor-guide/getting-started/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] ~Added/Updated related documentation.~
- [x] ~Added/Updated related tests.~

## Short description of the change(s)

<!--

If this is a non-trivial change, include information such as what
benefits this change brings as well as possible drawbacks.

If this fixes an existing issue, include the relevant issue number(s).

Thank you for taking the time to open this PR!

-->

Detects `yarn.lock` and uses `yarn` to install the missing hint packages in the `.hintrc` 

Resolves #1630

molant commented on packages/utils/src/npm.ts:
```diff
@@ -30,17 +30,25 @@ const install = (command: string) => {
     });
 };
 
+const hasYarnLock = (directory: string): Promise<boolean> => {
```
What about using [`promisify`](https://nodejs.org/dist/latest-v10.x/docs/api/util.html#util_util_promisify_original)  to do this?

```ts
import { promisify } from 'util';

const accessAsync = promisify(fs.access);

//...

packageManagerChoice = (await accessAsync(path.join(currentWorkingDir, 'yarn.lock'))) ? 'yarn' : 'npm';
```

shivangg replied:
@molant `promisify` returns Pomise<void> which can't be used to check for truthiness. 
![image](https://user-images.githubusercontent.com/10586637/57246941-287b9280-705c-11e9-9a9e-c236dcb3ad45.png)



Also, the same approach is used in `extension-browser` [here](https://github.com/webhintio/hint/blob/60624562af11362cf834f1791c6f3c1dfe84385d/packages/extension-vscode/src/server.ts#L29).

molant replied:
We have other things to worry about, let's leave it as it is 😊

molant commented on packages/utils/src/npm.ts:
```diff
@@ -113,17 +127,16 @@ const generateSearchQuery = (searchTerm: string, from?: number, size = 100) => {
  */
 export const search = async (searchTerm: string): Promise<NpmPackage[]> => {
-    const result = await npmRegistryFetch.json(generateSearchQuery(searchTerm)) as NpmSearchResults;
+    const result = (await npmRegistryFetch.json(generateSearchQuery(searchTerm))) as NpmSearchResults;
```
Do we need now the extra parems?

shivangg replied:
Added for readability (by Prettier). Should I remove those if they do not conform to style guide?

antross replied:
My vote is to leave them - I think having the extra parens here makes the intent clearer.

molant replied:
OK

molant commented:
@shivangg there are some failing tests. You can take a look in https://dev.azure.com/webhint/webhint/_build/results?buildId=3516&view=logs&jobId=da2b9a66-4613-5e8e-0da8-422d860ab222&taskId=a82dcf02-2c64-5cf1-7777-97ad98737900&lineStart=7282&lineEnd=7283&colStart=1&colEnd=1

antross commented on packages/utils/src/npm.ts:
```diff
@@ -62,10 +70,16 @@ export const installPackages = async (packages: string[]): Promise<boolean> => {
+        yarn: `yarn add${isDev ? ' --dev' : ''}`
+    };
+
+    const command: string = `${installCommand[packageManagerChoice]} ${packages.join(' ')}`;
```
```suggestion
    const command = `${installCommand[packageManagerChoice]} ${packages.join(' ')}`;
```

antross commented on packages/utils/tests/npm.ts:
```diff
@@ -255,7 +255,7 @@ test('installPackages should show the command to run if the installation fail an
+    t.true(childSpawnStub.args[0][0].includes('npm install --global hint1 @hint/formatter-formatter1'));
 });
 
 test('search should search the right query', async (t) => {
```
Let's add some new test cases in here covering the `yarn` case, both with and without `--dev`.

shivangg replied:
Added.

antross commented on packages/utils/src/npm.ts:
```diff
@@ -30,17 +30,25 @@ const install = (command: string) => {
 /** Install the given packages. */
 export const installPackages = async (packages: string[]): Promise<boolean> => {
     /** Whether or not the package should be installed as devDependencies. */
     let isDev: boolean = false;
```
Not technically part of your change, but I like to clean up unnecessary types when we're in the area.
```suggestion
    let isDev = false;
```

antross commented on packages/utils/src/npm.ts:
```diff
@@ -62,10 +70,16 @@ export const installPackages = async (packages: string[]): Promise<boolean> => {
             isDev = false;
         }
+
+        packageManagerChoice = (await hasYarnLock(currentWorkingDir)) ? 'yarn' : 'npm';
```
Global installs should use `npm`, regardless of if the project itself uses `yarn` (also a good thing to add a test case for).
```suggestion
        packageManagerChoice = (!global && await hasYarnLock(currentWorkingDir)) ? 'yarn' : 'npm';
```

shivangg replied:
@antross This statement is already in the `!global` block so `!global` check will be not required in the ternary operator.

antross replied:
Ah, I missed that. Perhaps moving setting `packageManagerChoice` outside of the `if (!global)` block would make this clearer? Putting it immediately after the `const global=` statement would also allow making this a `const packageManagerChoice = ...` declaration instead of a `let`.

shivangg replied:
Yeah, that would improve readability.

molant commented:
@shivangg https://github.com/webhintio/hint/pull/2327#discussion_r281317629 is still pending. I know you finish your internship today or tomorrow. Will you have time to finish this or should we take care of it?

Thanks!

shivangg commented:
@molant yes, I'll be working on to finish this.

molant commented on packages/utils/tests/npm.ts:
```diff
@@ -10,6 +10,8 @@ type Fs = {
     existsSync: () => boolean;
 };
 
+type HasYarnLock = () => boolean;
```
This should return a `Promise<boolean>`

shivangg replied:
@molant Changed these but `hasYarnLock` still not properly stubbed.

molant commented on packages/utils/tests/npm.ts:
```diff
@@ -148,6 +155,30 @@ test('installPackages should run the right command if `hint` is installed locall
+    const sandbox = t.context.sandbox;
+
+    sandbox.stub(t.context.fs, 'existsSync').returns(true);
+    sandbox.stub(t.context, 'hasYarnLock').returns(true);
```
Once you change the type it should be `.resolves(true)`

molant commented:
@shivangg looks like there are some issues with your PR?

[src/server.ts(20,10): error TS2305: Module '"../../../node_modules/@hint/utils/dist/src"' has no exported member 'hasYarnLock'.](https://dev.azure.com/webhint/webhint/_build/results?buildId=3795&view=logs&jobId=87b22c66-89ad-5aee-5190-c2f040f78e46&taskId=911263ff-bf20-50b2-0ed6-a32f919d9e58&lineStart=1976&lineEnd=1977&colStart=1&colEnd=1)

I think you need to update the reference there

antross commented on packages/utils/tests/npm.ts:
```diff
@@ -148,6 +155,30 @@ test('installPackages should run the right command if `hint` is installed locall
+    const sandbox = t.context.sandbox;
+
+    sandbox.stub(t.context.fs, 'existsSync').returns(true);
+    sandbox.stub(t.context, 'hasYarnLock').resolves(true);
```
```suggestion
    sandbox.stub(t.context, 'hasYarnLock').resolves(Promise.resolve(true));
```

shivangg replied:
`.resolves(Promise.resolve(true))` shows the error that argument of type 'Promise<boolean>' is not assignable to parameter of type 'boolean | undefined'

![image](https://user-images.githubusercontent.com/10586637/57412036-007f6100-720d-11e9-9dc7-abaf0c0592fd.png)


Did you mean `.returns(Promise.resolve(true))`? 
This doesn't show any error but fails to correctly stub `hasYarnLock`.

antross replied:
Yep, I read it too quickly and missed that `resolves` was already handling the `Promise` case (thought it was also `returns`).

shivangg commented:
@molant updated the reference to import `hasYarnLock` in `extension-vscode` correctly and now `extension-vscode` also has the same problem of stubbing `hasYarnLock`. Previously it was stubbing the `fs.access` which is not the case now.

molant commented:
> Previously it was stubbing the `fs.access` which is not the case now.

Can you update the tests for `extension-vscode`?

You'll have to update the following:

* This should be [`hasYarnLock`]( https://github.com/webhintio/hint/blob/f7c8314fae1f0962a546b56066e695a31288ddc8/packages/extension-vscode/tests/server.ts#L21) probably
* This should be the [`hasYarnLock` mock](https://github.com/webhintio/hint/blob/f7c8314fae1f0962a546b56066e695a31288ddc8/packages/extension-vscode/tests/fixtures/mocks.ts#L93). I think the mock fails so in this case you can probably do a `return Promise.resolve(false)`

shivangg commented:
Removed the `access` and extra `fs` (which also contained an `access` method) from the `extension-vscode` mocks. Added `hintUtils` that contains the stubbed method `hasYarnLock()`. All tests are passing now but an error is also thrown. 

<details>
<summary>Error</summary>

yarn test
yarn run v1.15.2
$ npm run lint && npm run build && npm run test-only

> vscode-webhint@1.0.5 lint /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> npm-run-all lint:*


> vscode-webhint@1.0.5 lint:js /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> eslint . --cache --ext js --ext md --ext ts --ignore-path ../../.eslintignore --report-unused-disable-directives


> vscode-webhint@1.0.5 lint:dependencies /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> node ../../scripts/lint-dependencies.js

Processing "/home/shivangg/Pesto/week9/hint/packages/extension-vscode"

> vscode-webhint@1.0.5 lint:md /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> node ../../scripts/lint-markdown.js


> vscode-webhint@1.0.5 build /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> npm-run-all build:*


> vscode-webhint@1.0.5 build:assets /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> cpx "./{src,tests}/**/{!(*.ts),.!(ts)}" dist


> vscode-webhint@1.0.5 build:ts /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> tsc -b


> vscode-webhint@1.0.5 test-only /home/shivangg/Pesto/week9/hint/packages/extension-vscode
> nyc ava


⠇ Error: Error
    at Object.fake.exceptionCreator (/home/shivangg/Pesto/week9/hint/node_modules/sinon/lib/sinon/default-behaviors.js:24:20)
    at Object.invoke (/home/shivangg/Pesto/week9/hint/node_modules/sinon/lib/sinon/behavior.js:150:35)
    at Object.functionStub (/home/shivangg/Pesto/week9/hint/node_modules/sinon/lib/sinon/stub.js:130:47)
    at Function.invoke (/home/shivangg/Pesto/week9/hint/node_modules/sinon/lib/sinon/spy.js:297:51)
    at Object.functionStub [as createAnalyzer] (/home/shivangg/Pesto/week9/hint/node_modules/sinon/lib/sinon/spy.js:152:22)
    at loadWebhint (/home/shivangg/Pesto/week9/hint/packages/extension-vscode/src/server.ts:32:44)
    at processTicksAndRejections (internal/process/task_queues.js:86:5)

  9 tests passed

------------------|----------|----------|----------|----------|-------------------|
File              |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------|----------|----------|----------|----------|-------------------|
All files         |    95.83 |     89.8 |    92.86 |    95.83 |                   |
 notifications.ts |      100 |      100 |      100 |      100 |                   |
 server.ts        |     95.8 |     89.8 |    92.86 |     95.8 |  54,55,91,158,224 |
------------------|----------|----------|----------|----------|-------------------|
Done in 9.00s.

</details>

Also, `utils` still has the test failing just like before because `hasYarnLock` not properly stubbed there.

molant commented:
I'll try to fix this over the weekend or early on Monday. Thanks @shivangg !

molant commented:
@shivangg the error is expected so that's fine. There was a problem with one of the tests in `@hint/utils` that I just fixed.
Hopefully CI passes and we can merge this 😊

Thanks!

shivangg commented:
@antross All tests are now passing. You can review now.

molant commented:
@shivangg I think @antross [requested some more tests](https://github.com/webhintio/hint/pull/2327/files#r281317629):

> Let's add some new test cases in here covering the `yarn` case, both with and without `--dev`.

shivangg commented:
@molant Oh yes, I had been waiting for 1 of them to pass before adding more of them. Will be adding them now.
